fix: resolve permission issues for non-root users

Freshwlnd opened the pull request:
# 问题描述 / Problem Description

当非 root 用户执行 `make` 命令时出现权限错误：

```bash
mv: 无法将"./logs" 移动至"./tmp/logs": 权限不够
failed to initialize build cache at /.cache/go-build: mkdir /.cache: permission denied
```

**根本原因**：
- Docker 容器内进程默认以 root 用户运行，创建的文件/目录归属 root
- Go 构建缓存路径权限问题，导致非 root 用户无法写入

# 解决方案 / Solution

## 1. Docker 用户权限映射
- 在 `Makefile` 中添加 `-u $(shell id -u):$(shell id -g)` 参数
- 让容器以当前用户身份运行，避免创建 root 拥有的文件

## 2. Go 构建缓存路径
- 设置 `GOCACHE=/workspace/.cache` 环境变量
- 确保 Go 构建缓存在有权限的工作目录下创建

## 3. 目录权限修复脚本
- 创建 `hack/ensure-directories.sh` 脚本
- 自动创建必要目录并修复权限问题
- 集成到 Makefile 构建流程中

# 变更内容 / Changes

- **Makefile**: 添加用户 ID 映射和 GOCACHE 环境变量设置
- **hack/ensure-directories.sh**: 新增权限修复脚本，确保目录正确ownership
- **构建流程**: 在 `default` 和 `serial-test` 目标中集成权限检查

Freshwlnd committed:
fix: resolve Docker container permission issues for non-root users

- Add user/group ID mapping to Docker containers to prevent root ownership
- Set GOCACHE to workspace directory to avoid permission denied errors
- Create ensure-directories.sh script to fix existing root-owned directories
- Integrate directory permission checks into Makefile targets

This fixes permission denied errors when running 'make' commands as non-root user,
ensuring all generated files and directories are owned by the current user.

Fixes issues where Docker containers create root-owned files/directories:
- ./logs, ./bin, ./gopath, ./registry-data directories
- Go build cache permission errors

Changes:
- Makefile: Add -u $(shell id -u):$(shell id -g) and GOCACHE env var
- hack/ensure-directories.sh: Script to ensure correct directory ownership
- Integrate ensure-directories target into build process

wzshiming commented on hack/ensure-directories.sh:
```diff
@@ -0,0 +1,55 @@
+    mkdir -p "$dir"
+    
+    # Check ownership
+    local owner=$(stat -c '%U' "$dir" 2>/dev/null || echo "none")
```
This doesn't work on MacOS.

```suggestion
    local owner=$(ls -ld "$dir" | awk '{print $3}' 2>/dev/null || echo "none")
```

Use ls + awk instead of stat for better compatibility

wzshiming commented on hack/ensure-directories.sh:
```diff
@@ -0,0 +1,55 @@
+    fi
+    
+    chmod 755 "$dir"
+    echo "Ensured directory: $dir (owner: $(stat -c '%U:%G' "$dir"))"
```
```suggestion
    echo "Ensured directory: $dir (owner: $(ls -ld "$dir" | awk '{print $3 ":" $4}'))"
```

wzshiming commented on Makefile:
```diff
@@ -41,8 +41,9 @@ GO_IMAGE ?= $(IMAGE_PREFIX)docker.io/library/golang:1.24
+	-u $(shell id -u):$(shell id -g) \
 	-v $(shell pwd):/workspace/ -w /workspace/ \
-	-e GOOS=$(GOOS) -e CGO_ENABLED=0 -e GOPATH=/workspace/gopath/ -e GOPROXY=$(GOPROXY) $(GO_IMAGE)
+	-e GOOS=$(GOOS) -e CGO_ENABLED=0 -e GOPATH=/workspace/gopath/ -e GOPROXY=$(GOPROXY) -e GOCACHE=/workspace/.cache $(GO_IMAGE)
```
```suggestion
	-e GOOS=$(GOOS) -e CGO_ENABLED=0 -e GOPATH=/workspace/gopath/ -e GOPROXY=$(GOPROXY) -e GOCACHE=/workspace/go-build $(GO_IMAGE)
```

Add the file to the .gitignore file to ignore it in Git.

wzshiming reviewed (commented):
Some small suggestions. Thank you

Freshwlnd committed:
Address review feedback: improve macOS compatibility and cache management

- Change GOCACHE from /workspace/.cache to /workspace/go-build for clearer naming
- Add go-build/ to .gitignore to exclude Go build cache from Git tracking
- Replace stat -c with ls -ld + awk for better macOS compatibility
- Add go-build directory to ensure-directories.sh script

This addresses the review comments from @wzshiming regarding:
1. macOS compatibility issues with stat command
2. Proper .gitignore configuration for build artifacts
3. More descriptive naming for Go build cache directory

Freshwlnd commented:
感谢您的审查意见！我已经完成了所有建议的修改：

## 1. macOS 兼容性修复
- 将 `stat -c` 命令替换为 `ls -ld + awk` 组合
- 确保脚本在 macOS 和 Linux 系统上都能正常工作

## 2. Go 构建缓存优化
- 将 GOCACHE 路径从 `/workspace/.cache` 改为 `/workspace/go-build`
- 添加 `go-build/` 到 `.gitignore` 文件，避免构建缓存被 Git 跟踪

## 3. 脚本完善
- 在 `ensure-directories.sh` 中添加 `go-build` 目录的权限检查
- 确保所有构建相关目录都有正确的所有权

修改后的代码已经过测试，在 Linux 系统非 root 用户环境下能够正常工作。
所有审查意见都已响应，请再次审查。谢谢！

wzshiming commented:
Thank you!
